[Lens] Color mapping for categorical dimensions

markov00 opened the pull request:
## Summary

This PR introduces the new color mapping feature into Lens.

The color mapping feature is introduced as a standalone sharable component available from `@kbn/coloring`. The [README.md](https://github.com/elastic/kibana/blob/ddd216457d66912de43c7688ae99044c2c34bbd2/packages/kbn-coloring/src/shared_components/color_mapping/README.md) file describes the components and the logic behind it.

The Color Mapping component is also connected to Lens and is available in the following charts:
- XY (you can specify the mappings from a breakdown dimension
- Partition (you can specify the mappings from the main slice/group by dimension)
- Tag cloud (you can specify the mappings from the tags dimension)

This MVP feature will be released under the Tech Preview flag.
This PR needs to prove the user experience and the ease of use. UI styles, design improvements and embellishments will be released in subsequent PRs. 

The current MVP-provided palettes are just a placeholder. I'm coordinating with @gvnmagni for a final set of palettes.

close https://github.com/elastic/kibana/issues/155037
close https://github.com/elastic/kibana/issues/6480
fix https://github.com/elastic/kibana/issues/28618
fix https://github.com/elastic/kibana/issues/96044
fix https://github.com/elastic/kibana/issues/101942
fix https://github.com/elastic/kibana/issues/112839
fix https://github.com/elastic/kibana/issues/116634


## Release note

This feature introduces the ability to change and map colors to break down dimensions in Lens. The feature provides an improved way to specify colors and their association with categories by giving the user a predefined set of color choices
or customized one that drives the user toward a correct color selection.
It provides ways to pick new colors and generate gradients. 
This feature is in Tech Preview and is enabled by default on every new visualization but can be turned off at will.
![image](https://github.com/elastic/kibana/assets/1421091/d03e59f8-4a6f-4761-ab4c-c53a57c1723a)


<!-- 
```[tasklist]
### Tasks
- [x] Extract the categories to automatically color from the visualization and pass them to the color assignment function
- [x] Take the color mapping helper function and apply it to the various visualizations
- [x] Use and apply the right theme
- [x] Use the new gradient design
- [x] Replace default `window.confirm` with EUI modals
- [x] Handle numeric fields as categorical and stringified where needed
- [x] Detect if the visualization is a new visualization or an existing one. Show new color mapping only on new charts
- [x] Move available palette into ito the coloring package
- [x] Add unit tests
- [x] Keep color mapping between chart types switches
- [x] Update existing functional tests to handle color mapping
- [x] Add i18n
- [x] telemetry counter for the legacy/new color mapping
- [x] Add README
- [x] Add functional tests
``` -->

jbudz reviewed (approved):
limits.yml

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

stratoula reviewed (commented):
This looks really fantastic, great work here Marco. I decided to play first with it and works perfectly except from some cases I mention below.

@MichaelMarcialis I think the technical preview can be a badge next to the switch, wdyt?

markov00 commented on packages/kbn-coloring/src/shared_components/color_mapping/README.md:
>When I assign a gradient color mapping and click the chart to create the filter, the series has the assignment to this series is lost. It displays the first color of the default palette although I haven't chosen this

Fixed in [1f8e9f7](https://github.com/elastic/kibana/pull/162389/commits/1f8e9f7c0b7fbfe396651f9da069a4a5e8c4d06f)

markov00 replied:
>Why my color mapping doesnt work in Filters?

The problem with color mappings and filters is the following: sometimes the filter creates “labels” an empty space in the end and the same empty space is kept also in the data query and response.
<img width="238" alt="Screenshot 2023-09-12 at 14 26 53" src="https://github.com/elastic/kibana/assets/1421091/62dafc60-3c29-49a2-8b28-dbc13ceb04b8">

In the color mapping, I tried to be a bit more "friendly" and I'm trimming whitespaces on the terms to avoid problems where the user accidentally adds a space at the end of a term and the rule match will not work.

I see two solutions:
- Remove color mapping white space trimming logic (the user should be carefully write the categories)
- Apply whitespace trimming logic to the filters (this will not work for pre-existing charts)

markov00 replied:
@stratoula this should solve the the whitespace trimming issue https://github.com/elastic/kibana/pull/162389/commits/2e876ab5158c88a152f105d38d7944d121fdd8a2

stratoula replied:
Problems mentioned here are resolved 👍 Thanx Marco

dej611 commented:
@markov00 I think `d3-scale` and `d3-<somethingelse>` added quite a bit of weight to some bundles (>100kb async to `lens`, `xy`, `tagCloud` and `partitionVis`):

XY:
<img width="622" alt="Screenshot 2023-09-12 at 16 38 22" src="https://github.com/elastic/kibana/assets/924948/169cd304-f848-45b1-9e35-d5fe5be7570d">

Lens:
<img width="386" alt="Screenshot 2023-09-12 at 16 29 47" src="https://github.com/elastic/kibana/assets/924948/1d95571d-b71e-4dce-8768-063c2fd9c7af">

To compare you can use the analysis from another PR I have up to date with bundle analysis:
Your PR: https://ci-artifacts.kibana.dev/webpack_bundle_analyzer/pr-162389/23d61c8d779871eb80563ef5d81db1c5c7ecdb53

~`main` (small tweak from `main`, but comparable): https://ci-artifacts.kibana.dev/webpack_bundle_analyzer/pr-159563/3f727f6c5245e18886fd149937adf2a170f06d5f

markov00 commented:
> @markov00 I think `d3-scale` and `d3-<somethingelse>` added quite a bit of weight to some bundles (>100kb async to `lens`, `xy`, `tagCloud` and `partitionVis`):

Thanks @dej611 I have changed the implementation from using d3 to use chroma-js here https://github.com/elastic/kibana/pull/162389/commits/04647eeb9baced50bbe4fda1ad7ea50c705cf7ee
let's see it this changes the situation.

dej611 commented:
> > @markov00 I think `d3-scale` and `d3-<somethingelse>` added quite a bit of weight to some bundles (>100kb async to `lens`, `xy`, `tagCloud` and `partitionVis`):
> 
> Thanks @dej611 I have changed the implementation from using d3 to use chroma-js here [04647ee](https://github.com/elastic/kibana/commit/04647eeb9baced50bbe4fda1ad7ea50c705cf7ee) let's see it this changes the situation.

Seems like redux is the remaining offender there with 60+ kb for 4 modules. Is that strictly required for this feature?
If it's important to keep it we could think to bundle `redux` within the shared bundle to shove off 60+ kb from multiple plugins in one go cc @elastic/kibana-core

markov00 commented:
> Seems like redux is the remaining offender there with 60+ kb for 4 modules. Is that strictly required for this feature? If it's important to keep it we could think to bundle `redux` within the shared bundle to shove off 60+ kb from multiple plugins in one go cc @elastic/kibana-core

The alternative is to refactor the code to use React hooks/state here and there, but I wasn't too confident in that. 
I believe that we can keep this 60k per bundle for now since this is an MVP, test it out and then reduce it in one way or the other. Up to @stratoula for a decision here

jughosta reviewed (commented):
Awesome work!

pgayvallet commented:
@dej611 I think it would make sense, but you need to check with @elastic/kibana-operations, they're the ones in charge of the shared bundles

stratoula commented:
> The alternative is to refactor the code to use React hooks/state here and there, but I wasn't too confident in that.
I believe that we can keep this 60k per bundle for now since this is an MVP, test it out and then reduce it in one way or the other. Up to @stratoula for a decision here

The bundle size got reduced in half so I am already happy with it. Let's not do any refactoring now for 60K. @dej611 can you discuss your proposal with the operations team?

stratoula commented on src/plugins/data/common/search/aggs/buckets/multi_field_key.ts:
```diff
@@ -38,3 +38,13 @@ export class MultiFieldKey {
+  return field instanceof MultiFieldKey;
+}
+
+/**
+ * Multi-field key separator used in Visualizations (Lens, AggBased, TSVB).
+ * This differs from the separator used in the toString method of the MultiFieldKey
+ */
+export const MULTI_FIELD_KEY_SEPARATOR = ' › ';
```
I think that this is the multi fields aggregation field so it makes sense to be on the data plugin, right @markov00 ?

dej611 replied:
Correct. We need a single place to put the separator, and the data makes more sense as it is a dependency of all visualization plugins.

jughosta replied:
Thanks! Would a name like `CATEGORIES_MULTI_FIELD_KEY_SEPARATOR` be more descriptive about its application?

jughosta replied:
Having a separate package would be great! Also from codeowners-perspective so you don't have to wait for reviews from Data Discovery team for visualizations related changes 🙂

markov00 replied:
comment added to https://github.com/elastic/kibana/issues/138723

jughosta reviewed (approved):
Data Discovery changes LGTM 👍

drewdaemon reviewed (changes requested):
Overall, amazing work. I think it's a really great feature and I personally love the clean architecture.

Multi-metric partition chart coloring is currently broken on this PR, so probably shouldn't merge until that's taken care of (requesting changes to make this clear despite the other approvals).

https://github.com/elastic/kibana/assets/315764/33b8677e-51dc-4e2f-a10d-713d8fcf6ee5

This also made me realize that I never added functional test for that feature. Sorry about that—it's kind of obscure so easy to break. Created [an issue](https://github.com/elastic/kibana/issues/166497).

One UX issue I noticed. When the flyout opens downwards, the custom color picker can become lost below the fold. Maybe have the flyout open upwards no matter what?

https://github.com/elastic/kibana/assets/315764/946a7958-ed54-4967-91aa-e05786c336bc

A few code questions, but nothing blocking.

markov00 commented on packages/kbn-coloring/src/shared_components/color_mapping/color/rule_matching.ts:
```diff
@@ -0,0 +1,46 @@
+        );
+      }
+      return rule.values.includes(`${value}`);
+    case 'matchExactlyCI':
```
Yep, I know, when implementing the component I made space and some test by adding also these possibilities (you can see that we also have the range rule, even if we are not going to expose them soon.
Except for the range, that will be part of the next steps, I can remove the other Rules configuration

drewdaemon replied:
If it's just a possibility, I'd vote to take it out. But, if it's definitely planned, I'm happy with leaving it in

markov00 replied:
together with @stratoula we decided to keep this in the code for now

markov00 commented on packages/kbn-coloring/src/shared_components/color_mapping/state/selectors.ts:
```diff
@@ -0,0 +1,26 @@
+import { getPalette } from '../palette';
+import { RootState } from './color_mapping';
+
+export function getPaletteSelector(getPaletteFn: ReturnType<typeof getPalette>) {
+  return (state: RootState) => getPaletteFn(state.colorMapping.paletteId);
+}
+export function getColorModeSelector(state: RootState) {
+  return state.colorMapping.colorMode;
+}
+export function getSpecialAssignmentsSelector(state: RootState) {
+  return state.colorMapping.specialAssignments;
+}
+export function isAutoAssignmentModeSelector(state: RootState) {
+  return state.colorMapping.assignmentMode === 'auto';
+}
+export function getColorPickerVisibilitySelector(state: RootState) {
+  return state.ui.colorPicker;
+}
```
oh, sorry, that was the convention used in elastic-charts. I can change that

markov00 replied:
renamed here https://github.com/elastic/kibana/pull/162389/commits/0e7f29d3cdbed05ee570473ebdd17f3ac72f1024

markov00 commented on packages/kbn-coloring/src/shared_components/color_mapping/components/color_picker/rgb_picker.tsx:
```diff
@@ -0,0 +1,111 @@
+import { getColorContrast } from '../../color/color_math';
+import { getPalette } from '../../palette';
+
+export function RGBPicker({
+  isDarkMode,
+  color,
+  getPaletteFn,
+  selectColor,
+  close,
+}: {
```
I don't believe is necessary now. The idea here is to improve this further post MVP, so I prefer to spend time later on unifying it.

markov00 commented on packages/kbn-coloring/src/shared_components/color_mapping/README.md:
```diff
@@ -0,0 +1,84 @@
+This package provides:
+- a React component, called `CategoricalColorMapping` that provides a simplified UI (that in general can be hosted in a flyout), that helps the user generate a `ColorMapping.Config` object that descibes the mappings configuration
+- a function `getColorFactory` that given a color mapping configuration returns a function that maps a passed category to the corresponding color
+- a definition scheme for the color mapping, based on the type `ColorMapping.Config`, that provides an extensible way of describing the link betwe
```
sorry you are right, I've finished the README here https://github.com/elastic/kibana/pull/162389/commits/385fca482d2d3f8c53443f5f9c60d8eb0747c688

markov00 commented:
@drewdaemon thanks for the review

> Multi-metric partition chart coloring is currently broken on this PR, so probably shouldn't merge until that's taken care of (~requesting changes to make this clear despite the other visualizations approvals~ Edit: I thought someone from the team had already approved... I was wrong so count this a comment review 😅 ).

Fixed here https://github.com/elastic/kibana/pull/162389/commits/6ebd87729e77e7d41754caf7012bd53b0c0da1d2



> One UX issue I noticed. When the flyout opens downwards, the custom color picker can become lost below the fold. Maybe have the flyout open upwards no matter what?


I've fixed that here https://github.com/elastic/kibana/pull/162389/commits/ae2fe91a6f3909236f08231292e33006ed9482ef and also moved the opening a bit to the right, so we don't cover both the chart and the flyout. I've also added the ability to scroll relative to the anchored element that helps when inevitably the popover overflows https://github.com/elastic/kibana/pull/162389/commits/3ec9a028e23837aa6bb889497a7805c2c55ef7af

drewdaemon commented:
I noticed that the Lens editor occasionally crashes with the following messages

<img width="1423" alt="Screenshot 2023-09-18 at 6 29 29 AM" src="https://github.com/elastic/kibana/assets/315764/9c0942c3-986c-40c3-9f2e-0f1e58c4420b">

I haven't yet figured out a reliable set of steps to reproduce.

stratoula commented:
@markov00 

- the bugs that I mentioned above are all fixed 👍 
- the bug that Drew mentioned on multiple metrics is fixed. The only thing that doesn't work is that if I add a slice by dimension and change the palette, the changes are not respected
<img width="1491" alt="image" src="https://github.com/elastic/kibana/assets/17003240/4d9d9a61-7677-4e21-9d33-73372d5c1fba">

- about that:

> I noticed that the Lens editor occasionally crashes with the following messages

I can't replicate it

Except from the bug I mention above I can't find anything else 🎉

markov00 commented:
@stratoula 
> * the bug that Drew mentioned on multiple metrics is fixed. The only thing that doesn't work is that if I add a slice by dimension and change the palette, the changes are not respected

I've fixed that in https://github.com/elastic/kibana/pull/162389/commits/2b889e8a2a3ad58c81b0e741c2a538d00102a43d
please take a look.
You may notice some issues with previously saved color mapping because I've changed the palette ids in accordance with @gvnmagni

stratoula reviewed (approved):
Changes LGTM, I tested the last bug with multiple metrics on partition charts. Everything works as expected 👍 

The only thing that I think needs improvement is the technical preview text but I am pretty sure that @MichaelMarcialis will figure it out. I am approving as I feel that it works great. 💯

MichaelMarcialis reviewed (commented):
Excellent work all around, Marco! I'm so impressed with how quickly you've pulled this together. I've given the PR a test drive and have listed my initial comments below for your review. Per your request, I added it as a single comment on the `README.md` file. Let me know if you have any questions.

MichaelMarcialis commented:
> @MichaelMarcialis I think the technical preview can be a badge next to the switch, wdyt?

@stratoula: Generally speaking, I agree. Though I do have some questions/concerns about the current scope and placement of that switch (as noted in my [review comments](https://github.com/elastic/kibana/pull/162389#discussion_r1332104911)) which may have implications.

stratoula commented:
The only thing I found is that in tagcloud the suggestions are not colored correctly (I consider it a nit)

<img width="1493" alt="image" src="https://github.com/elastic/kibana/assets/17003240/e6413897-9786-4941-aae2-d7903274fdda">

I tested everything else and seems fine.

markov00 commented:
> The only thing I found is that in tagcloud the suggestions are not colored correctly (I consider it a nit)

thanks @stratoula nice catch, I just fixed that and also aligned the colors set shown in the dimension button with the color mapping selected

stratoula reviewed (approved):
Tested again and it works fine. 🎉 

@MichaelMarcialis can you please review again? Marco has made the majority of the changes you asked. This PR introduces changes in many files so I would love this be merged before the FF 🙏

markov00 commented:
@elasticmachine merge upstream

Danouchka commented:
Hi @stratoula  do you think this feature has good chance to be released in 8.11 ?

stratoula commented:
@Danouchka yes we are very positive that this is going to be merged soonish :)
